Python/Product/Uwp.Interpreter/PythonUwpInterpreterFactoryProvider.cs

```diff
@@ -21,106 +21,381 @@
-                                                InterpreterUIMode.CannotBeDefault | InterpreterUIMode.SupportsDatabase
-                                            )
-                                        );
-                                    }
-                                }
+        private void ProviderValue_ProjectsChanaged(object sender, EventArgs e) {
+            var contextProvider = (IProjectContextProvider)sender;
+            if (contextProvider != null) {
+                var hasInterpreterFactoriesChanged = false;
+                var seen = new HashSet<string>();
+                var removed = new HashSet<ProjectInfo>();
+                lock (_projects) {
+
+                    foreach (var context in contextProvider.Projects) {
+                        var projectInfo = ProjectInfo.CreateFromProjectContext(context, contextProvider);
+                        if (projectInfo != null && projectInfo.FullPath != null) {
+                            seen.Add(projectInfo.FullPath);
+                            if (!_projects.ContainsKey(projectInfo.FullPath)) {
+                                _projects[projectInfo.FullPath] = projectInfo;
+                            } else {
+                                // reuse the old existing project info
+                                projectInfo.Dispose();
+                                projectInfo = _projects[projectInfo.FullPath];
                             }
+                            hasInterpreterFactoriesChanged |= projectInfo.DiscoverInterpreters();
                         }
                     }
+
+                    // Then remove any existing projects that are no longer there
+                    var toRemove = _projects
+                        .Where(x => x.Value.ContextProvider == contextProvider && !seen.Contains(x.Key))
+                        .Select(x => x.Key)
+                        .ToArray();
+
+                    foreach (var proj in toRemove) {
+                        var projInfo = _projects[proj];
+                        _projects.Remove(proj);
+                        if (projInfo.Factory != null) {
+                            hasInterpreterFactoriesChanged = true;
+                        }
+                        projInfo.Dispose();
+                    }
+                }
+
+                if (hasInterpreterFactoriesChanged) {
+                    OnInterpreterFactoriesChanged();
                 }
-            } catch (IOException) {
-                // IOException is not critical here, just means we cannot interrogate for factories at this point
+            }
+        }
+
+        /// <summary>
+        /// Represents an MSBuild project file.  The file could have either been read from
+        /// disk or it could be a project file running inside of the IDE which is being
+        /// used for a Python project node.
+        /// </summary>
+        private class MSBuildProjectInfo : ProjectInfo {
```

Comment:
There shouldn't be any need to clone these classes as you'll be getting the context provider exported by the main implementation.